fuel-p2p/src/peer_info.rs

```diff
@@ -0,0 +1,384 @@
+            peer_id,
+            connection_id,
+            connected_point,
+            failed_addresses,
+        );
+
+        self.identify.inject_connection_established(
+            peer_id,
+            connection_id,
+            connected_point,
+            failed_addresses,
+        );
+
+        self.insert_peer(peer_id, connected_point.clone());
+    }
+
+    fn inject_connected(&mut self, peer_id: &PeerId) {
+        let addresses = self.addresses_of_peer(peer_id);
+        self.insert_peer_addresses(peer_id, addresses);
+        self.ping.inject_connected(peer_id);
+        self.identify.inject_connected(peer_id);
+    }
+
+    fn inject_disconnected(&mut self, peer_id: &PeerId) {
+        self.ping.inject_disconnected(peer_id);
+        self.identify.inject_disconnected(peer_id);
+
+        // todo: we could keep it in a cache for a while
+        self.peers.remove(peer_id);
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        loop {
+            match self.ping.poll(cx, params) {
+                Poll::Pending => break,
+                Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+                    peer_id,
+                    handler,
+                    event,
+                }) => {
+                    return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+                        peer_id,
+                        handler,
+                        event: EitherOutput::First(event),
+                    })
+                }
+                Poll::Ready(NetworkBehaviourAction::ReportObservedAddr { address, score }) => {
+                    return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+                        address,
+                        score,
+                    })
+                }
+                Poll::Ready(NetworkBehaviourAction::CloseConnection {
+                    peer_id,
+                    connection,
+                }) => {
+                    return Poll::Ready(NetworkBehaviourAction::CloseConnection {
+                        peer_id,
+                        connection,
+                    })
+                }
+                Poll::Ready(NetworkBehaviourAction::Dial { handler, opts }) => {
+                    let handler =
+                        IntoProtocolsHandler::select(handler, self.identify.new_handler());
+
+                    return Poll::Ready(NetworkBehaviourAction::Dial { handler, opts });
+                }
+                Poll::Ready(NetworkBehaviourAction::GenerateEvent(ping_event)) => {
+                    if let PingEvent {
+                        peer,
+                        result: Ok(PingSuccess::Ping { rtt }),
+                    } = ping_event
+                    {
+                        self.insert_latest_ping(&peer, rtt)
+                    }
+                }
```

Comment:
the if-let can be combined with the match statement, ie

```suggestion
                Poll::Ready(NetworkBehaviourAction::GenerateEvent(PingEvent {
                        peer,
                        result: Ok(PingSuccess::Ping { rtt }),
                })) => {
                    self.insert_latest_ping(&peer, rtt)
                }
```